### ISSUE TITLE
[Feature Request] Have `unifunc cluster_function` default read from stdin if no input is provided

### ISSUE THREAD

**Issue by jolespin** [PedroMTQ/UniFunc#2]
I'm working on implementing this into some pipelines and an option for reading from stdin would be really useful. 

```
    def read_clustered_annotations(self):
        #this yields all clusters, one by one - less memory footprint
        #FILE NEEDS TO BE SORTED BY CLUSTER_ID
        if self.input_path is None:
                file = sys.stdin
        else:
                file = open(self.input_path)
        
            temp=[]
        
            for line in file.readlines():
                line=line.strip()
                line=line.split('\t')
                gene_id,cluster_id,annotation=line
                annotation=self.pre_process_annotations(annotation)
                if temp and cluster_id!=previous_cluster_id:
                    yield previous_cluster_id,temp
                    temp=[]
                temp.append([gene_id,annotation])
                previous_cluster_id=cluster_id
        if temp:
            yield previous_cluster_id,temp
        if file is not sys.stdin:
                file.close()
```

What are your thoughts?

**Comment by PedroMTQ** [comment]
Hey,

Yes, this can be added. You just need to make sure the stdin is sorted by cluster id, because the function yield depends on that.
Alternatively I could change it to be independent of file order, but this would be less memory efficient.

I attached the `Cluster_Representative_Function.py` as a zip. could you try it locally to see if it fits your requiremens?
I'll then package it and release it through conda.

(Btw, just a small tip, I would avoid using readlines() as this reads and stores the whole file in RAM. You can just iterate over the file with `for line in file`)

Regards,
[test.zip](https://github.com/PedroMTQ/UniFunc/files/7868980/test.zip)

Pedro

**Comment by jolespin** [comment]
Thanks for the tip on readlines. I m not sure why I did readlines there instead of looping the file.

I'm wondering if it might be better to just read in the cluster mapping and sort even though it might be a tad slower and more memory intensive. Since there's only 3 columns, the file would have to be pretty large to cause memory issues (not sure about the backend nltk models).  IMO trusting people to read the fine print on documentation or making one extra step for preprocessing will deter people from using it. 

Reading in as a pandas data frame will make the sorting easier and it supports stdin natively. 

Is pandas a dependency on UniFunc already?

**Comment by PedroMTQ** [comment]
Hello @jolespin 
Yes, you're right. I'll make it more agnostic.
Sorting is only necessary if the function is yielding the lines directly, otherwise I can just store it in a dictionary and yield it from there.

I've added the changes you proposed.
Would you like me to add/change anything else before I release the new version?

Regards,
Pedro

**Comment by jolespin** [comment]
Nothing else that I can think of that’s major off the top of my head. One thing I noticed IIRC is that the output directory requires a slash at the end or else it appends to that string. I’m not sure if this is the expected functionality or not as I’ve seen this used as basename in other programs like STAR. If that’s not the expected functionality then I would recommend doing something like os.path.join for the output files. Not a huge deal either way just mentioning it in case.

**Comment by PedroMTQ** [comment]
@jolespin 

I fixed that in the meantime. I'll create a release then.

Regards,
Pedro

**Comment by PedroMTQ** [comment]
Hello @jolespin 
The new release is available in conda now (1.4.1)
I'll close this issue, but feel free to open a new one if you run into any issues or need anything else.

Regards,
Pedro
